### PR TITLE
Initialize all embind bindings on all threads/workers

### PR DESCRIPTION
**Pull request by sbc100**
This change replaces `__embind_register_native_and_builtin_types` with
`_embind_register_bindings` which calls all `EMSCRIPTEN_BINDINGS`
functions.

Fixes: #17372

### CODE REVIEW

**Comment by sbc100** [comment]
Just a work in progress.. needs some tests..

File: system/lib/embind/bind.cpp [hunk 1]
```diff
@@ -99,13 +114,8 @@ template <typename T> static void register_memory_view(const char* name) {
...
 
-extern "C" {
-
-// Normal initialization, executed through a global constructor. This
-// happens on the main thread; pthreads will call it manually, so make
-// sure we also export it (via EMSCRIPTEN_KEEPALIVE).
-EMSCRIPTEN_KEEPALIVE __attribute__((constructor))
-void __embind_register_native_and_builtin_types() {
+EMSCRIPTEN_BINDINGS(builtin) {
+  //_emscripten_err("running core bindings");
```
Comment (brendandahl) [inline system/lib/embind/bind.cpp:118]: Leftover debugging?

**Review by brendandahl — APPROVED** [review]

**Comment by koushikchalla** [comment]
@sbc100 
Is there a way to disable running embind bindings on all threads? Also is there a way i can guarantee embind bindings run after main function is called. 

Reason for this is that the bindings i need are dynamic. There is a function called in main that fills a map of function name to function address (map<string,void (*)(string)>();) and then i bind by looping through this map

**Comment by sbc100** [comment]
> @sbc100 Is there a way to disable running embind bindings on all threads? Also is there a way i can guarantee embind bindings run after main function is called.

I don't think so.  Why are you wanting to that?  (Remember the cost of binding generation is per-worker, not per-pthread, so unless you binding code is super expensive its cost will be dwarfed by worker-creation time).

> 
> Reason for this is that the bindings i need are dynamic. There is a function called in main that fills a map of function name to function address (map<string,void (*)(string)>();) and then i bind by looping through this map

IIUC, the binding are always run **before** the main function is called because they are part of static construction.   Trying to change that behaviour would be tricky I think.   Perhaps there is some way you can fill in your map of function names at static constructor time instead?

**Comment by koushikchalla** [comment]
Understood. I'll put the map filling code in a static constructor. that should solve the problem

**Pull request merged**